core/reactor_common.c

```diff
@@ -1707,87 +1737,104 @@ void initialize_global(void) {
     // Federation trace object must be set before `initialize_trigger_objects` is called because it
     //  uses tracing functionality depending on that pointer being set.
     #ifdef FEDERATED
-    set_federation_trace_object(envs->trace);
+    lf_set_federation_trace_object(envs->trace);
     #endif
     // Call the code-generated function to initialize all actions, timers, and ports
     // This is done for all environments/enclaves at the same time.
     _lf_initialize_trigger_objects() ;
 }
 
+/** Flag to prevent termination function from executing twice. */
+bool _lf_termination_executed = false;
```

Comment:
I think that if the termination function can only be executed from the `atexit` hook, we should not need this global variable. Note that functions registered using `atexit` are executed whether we exit using `exit` or we return from `main`.

Reply:
I've updated the comment (commit #f4ab3d88cad18f16559f05d01fe7f7ae74bf6667) to indicate that this is also used to exit from potentially infinite loops in background threads.  I'm not sure whether this really makes any difference, and it might even be platform dependent, but I think it's better to be extra cautious here.  I was seeing quite a few problems arising only during abnormal termination in federated execution (e.g. zombie processes), and I am seeing fewer now (but possibly just because I'm seeing fewer abnormal terminations).